main.py

```diff
@@ -1,68 +1,225 @@
+from telegram.ext.filters import Filters
+from telegram.message import Message
+from telegram.utils.helpers import escape_markdown
+import database
+
+
+private_chat_id = -1001218939335
+private_chat_username = 'devs_chat'
+
+print('loading/creating database')
+DB_PATH = 'memebot.db'
+db = database.UserDB(DB_PATH)
+
+def save_db():
+	global db
+	if db.changed:
+		print('saving database')
+		db.changed = False
+		db.dump()
+
+class SaverThread(Thread):
+	'''
+	Class silimar to threading.Timer, but where Timer runs only once, this class runs function multiple times
+
+	To stop the loop, use function set on passed event: event.set()
+	'''
+
+	def __init__(self, event: Event, interval: float):
+		Thread.__init__(self)
+		self.stopped = event
+		self.interval = interval
+
+	def run(self):
+		while not self.stopped.wait(self.interval):
+			save_db()
+
+print('creating saving timer')
+SAVE_INTERVAL = 5 * 60 # 5 minutes
+stop_saving_flag = Event()
+save_timer = SaverThread(stop_saving_flag, SAVE_INTERVAL)
+save_timer.start()
+
+print("reading config")
+CURDIR=path.dirname(argv[0])
+with open(path.join(CURDIR,"./config.json"),"r") as f:
+	CONFIG=json.load(f)
+
+print("initializing commands")
+
+updater=Updater(CONFIG["TOKEN"])
+
+def command(name:str):#some python magic
+	def add_it(func):
+		updater.dispatcher.add_handler(CommandHandler(name, func))
+	return add_it
+
+def message(filters):
+	def add_it(func):
+		updater.dispatcher.add_handler(MessageHandler(filters,func))
+	return add_it
+
+def filter_chat(id, chat):
+	'''
+	id: id of a chat
+	chat: @<chat> without @
+	'''
+	def decorator(function):
+		def wrapper(update: Update, context: CallbackContext):
+			if update.message.chat_id != id:
+				update.message.chat.send_message(f'This feature only works in chat @{chat}')
+				return
+			return function(update, context)
+		return wrapper
+	return decorator
+
+
+@command("ping")
+def ping(update: Update, context: CallbackContext):
+	dt=datetime.now(update.message.date.tzinfo)-update.message.date
+	update.message.reply_text(f'Ping is {dt.total_seconds():.2f}s')
+
+def escape_md(txt: str) -> str:
+	return escape_markdown(txt, 2)
+
+def get_mention(user: User):
+	return user.mention_markdown_v2()
+
+@message(Filters.status_update.new_chat_members)
+def new_chat_member(update: Update, context: CallbackContext):
+	handles=", ".join(get_mention(user) for user in update.message.new_chat_members)
+	update.message.reply_text(
+f"""{handles},
+いらっしゃいませ\\! \\[Welcome\\!\\]
+Welcome to this chat\\! Please read the rules\\.
+Добро пожаловать в чат\\! Прочти правила, пожалуйста\\.
+このチャットへようこそ！ ルールをお読みください。
+
+[rules](https://t\\.me/dev\\_meme/3667)""",
+parse_mode=ParseMode.MARKDOWN_V2)
+
+def is_admin(chat: Chat, user: User) -> bool:
+	# might wanna cache admins
+	status = chat.get_member(user.id).status
+	return status == 'creator' or status == 'administrator'
+
+def get_warn_target(message: Message) -> User | None:

```

Comment:
rename to `get_reply_target` or something to generalize the function - also don't check if .from_user is `None`, since you'd return `None` in that case anyway.

Reply:
Yes, good change. It was a function that returned multiple targets, but it's too hard to implement